File: src/main/scala/jk_5/nailed/server/world/NailedWorld.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.world

import java.util

import jk_5.nailed.api.gamerule.{DefaultGameRules, EditableGameRules}
import jk_5.nailed.api.map.Map
import jk_5.nailed.api.mappack.metadata.MappackWorld
import jk_5.nailed.api.player.Player
import jk_5.nailed.api.world._
import jk_5.nailed.server.map.gamerule.WrappedEditableGameRules
import jk_5.nailed.server.player.NailedPlayer
import net.minecraft.network.play.server.S41PacketServerDifficulty
import net.minecraft.world.{EnumDifficulty, WorldServer}

import scala.collection.mutable

/**
 * No description given
 *
 * @author jk-5
 */
class NailedWorld(var wrapped: WorldServer, val context: WorldContext = null) extends World {

  private var map: Option[Map] = None
  private val playerSet = mutable.HashSet[Player]()
  private var players = new Array[Player](0)

  private val provider: Option[WorldProvider] = wrapped.provider match {
    case p: DelegatingWorldProvider => Some(p.wrapped)
    case _ => None
  }

  this.getConfig match {
    case c: MappackWorld => setDifficulty(c.difficulty())
    case _ =>
  }

  override val getGameRules: EditableGameRules = if(this.context != null && this.context.getConfig != null) new WrappedEditableGameRules(context.getConfig.gameRules) else new WrappedEditableGameRules(DefaultGameRules.INSTANCE)

  override def getDimensionId = wrapped.provider.getDimensionId
  override def getName = "world_" + getDimensionId
  override def getPlayers: util.Collection[Player] = util.Arrays.asList(players: _*)
  override def getDimension = this.provider match {
    case Some(p) => p.getDimension
    case None => Dimension.OVERWORLD
  }

  override def setMap(map: Map) = this.map = Some(map)
  override def getMap = this.map.orNull
  override def getConfig: MappackWorld = if(this.context == null) null else this.context.getConfig

  override def onPlayerJoined(player: Player){
    println("Player " + player.toString + " joined world " + this.toString)
    playerSet += player
    players = playerSet.toArray
    player.asInstanceOf[NailedPlayer].sendPacket(new S41PacketServerDifficulty(wrapped.getDifficulty, false))
    if(this.getConfig.resourcePackUrl() != null){
      player.loadResourcePack(this.getConfig.resourcePackUrl(), "") //TODO: fix hash
    }
  }

  override def onPlayerLeft(player: Player){
    println("Player " + player.toString + " left world " + this.toString)
    playerSet -= player
    players = playerSet.toArray
  }

  override def getTime = this.wrapped.getWorldTime.toInt
  override def setTime(time: Int) = this.wrapped.setWorldTime(time)

  override def getWeather: WeatherType = {
    val rain = this.wrapped.isRaining
    val thunder = this.wrapped.isThundering

    if(!rain && !thunder) WeatherType.DRY
    else if(!thunder) WeatherType.RAIN
    else WeatherType.THUNDER
  }
  override def setWeather(weather: WeatherType) = weather match {
    case WeatherType.DRY =>
      this.wrapped.getWorldInfo.setRaining(false)
      this.wrapped.getWorldInfo.setRainTime(0)
      this.wrapped.getWorldInfo.setThundering(false)
      this.wrapped.getWorldInfo.setThunderTime(0)
    case WeatherType.RAIN =>
      this.wrapped.getWorldInfo.setRaining(true)
      this.wrapped.getWorldInfo.setThundering(false)
      this.wrapped.getWorldInfo.setThunderTime(0)
    case WeatherType.THUNDER =>
      this.wrapped.getWorldInfo.setRaining(true)
      this.wrapped.getWorldInfo.setThundering(true)
  }

  override def getDifficulty = Difficulty.byId(wrapped.getDifficulty.getDifficultyId)
  override def setDifficulty(difficulty: Difficulty){
    val diff = EnumDifficulty.getDifficultyEnum(difficulty.getId)
    this.wrapped.getWorldInfo.setDifficulty(diff)
    this.players.map(_.asInstanceOf[NailedPlayer]).foreach(_.sendPacket(new S41PacketServerDifficulty(diff, false)))
    difficulty match {
      case Difficulty.HARD | Difficulty.NORMAL | Difficulty.EASY => wrapped.setAllowedSpawnTypes(true, true)
      case Difficulty.PEACEFUL => wrapped.setAllowedSpawnTypes(false, true)
    }
  }

  override def toString = s"NailedWorld{id=$getDimensionId,name=$getName,dimension=$getDimension,gameRules=$getGameRules}"
}

File: src/main/java/jk_5/nailed/server/world/NailedWorld.java
package jk_5.nailed.server.world;

import com.google.common.base.MoreObjects;
import com.google.common.collect.ImmutableSet;
import jk_5.nailed.api.gamerule.DefaultGameRules;
import jk_5.nailed.api.gamerule.EditableGameRules;
import jk_5.nailed.api.map.Map;
import jk_5.nailed.api.mappack.metadata.MappackWorld;
import jk_5.nailed.api.player.Player;
import jk_5.nailed.api.world.*;
import jk_5.nailed.server.map.gamerule.WrappedEditableGameRules;
import jk_5.nailed.server.player.NailedPlayer;
import net.minecraft.network.play.server.S41PacketServerDifficulty;
import net.minecraft.world.EnumDifficulty;
import net.minecraft.world.WorldServer;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

public class NailedWorld implements World {

    private static final Logger logger = LogManager.getLogger();

    private final Set<Player> players = new HashSet<Player>();
    private final WorldContext context;
    private final WorldProvider worldProvider;
    private final EditableGameRules gameRules;

    private WorldServer wrapped;
    private Map map;

    public NailedWorld(WorldServer wrapped, WorldContext context) {
        this.wrapped = wrapped;
        this.context = context;

        if(wrapped.provider instanceof DelegatingWorldProvider){
            this.worldProvider = ((DelegatingWorldProvider) wrapped.provider).getWrapped();
        }else{
            this.worldProvider = null;
        }

        MappackWorld config = this.getConfig();
        if(config != null){
            this.setDifficulty(config.difficulty());
        }

        if(this.context != null && this.context.getConfig() != null){
            this.gameRules = new WrappedEditableGameRules(context.getConfig().gameRules());
        }else{
            this.gameRules = new WrappedEditableGameRules(DefaultGameRules.INSTANCE);
        }
    }

    @Override
    public EditableGameRules getGameRules() {
        return this.gameRules;
    }

    @Override
    public int getDimensionId() {
        return wrapped.provider.getDimensionId();
    }

    @Override
    public String getName() {
        return "world_" + getDimensionId();
    }

    @Override
    public Collection<Player> getPlayers() {
        return ImmutableSet.copyOf(players);
    }

    @Override
    public Dimension getDimension() {
        if(this.worldProvider != null){
            return this.worldProvider.getDimension();
        }else{
            return Dimension.OVERWORLD;
        }
    }

    @Override
    public void setMap(Map map) {
        this.map = map;
    }

    @Override
    public Map getMap() {
        return this.map;
    }

    @Override
    public MappackWorld getConfig() {
        if(this.context == null){
            return null;
        }
        return this.context.getConfig();
    }

    @Override
    public void onPlayerJoined(Player player) {
        logger.info("Player " + player.toString() + " joined world " + this.toString());
        players.add(player);
        ((NailedPlayer) player).sendPacket(new S41PacketServerDifficulty(wrapped.getDifficulty(), false));
        if(this.getConfig().resourcePackUrl() != null){
            player.loadResourcePack(this.getConfig().resourcePackUrl(), ""); //TODO: fix hash
        }
    }

    @Override
    public void onPlayerLeft(Player player) {
        logger.info("Player " + player.toString() + " left world " + this.toString());
        players.remove(player);
    }

    @Override
    public int getTime() {
        return (int) wrapped.getWorldTime();
    }

    @Override
    public void setTime(int time) {
        this.wrapped.setWorldTime(time);
    }

    @Override
    public WeatherType getWeather() {
        boolean rain = this.wrapped.isRaining();
        boolean thunder = this.wrapped.isThundering();

        if(!rain && !thunder){
            return WeatherType.DRY;
        }else if(!thunder){
            return WeatherType.RAIN;
        }else{
            return WeatherType.THUNDER;
        }
    }

    @Override
    public void setWeather(WeatherType weather) {
        if(weather == WeatherType.DRY){
            this.wrapped.getWorldInfo().setRaining(false);
            this.wrapped.getWorldInfo().setRainTime(0);
            this.wrapped.getWorldInfo().setThundering(false);
            this.wrapped.getWorldInfo().setThunderTime(0);
        }else if(weather == WeatherType.RAIN){
            this.wrapped.getWorldInfo().setRaining(true);
            this.wrapped.getWorldInfo().setThundering(false);
            this.wrapped.getWorldInfo().setThunderTime(0);
        }else if(weather == WeatherType.THUNDER){
            this.wrapped.getWorldInfo().setRaining(true);
            this.wrapped.getWorldInfo().setThundering(true);
        }
    }

    @Override
    public Difficulty getDifficulty() {
        return Difficulty.byId(wrapped.getDifficulty().getDifficultyId());
    }

    @Override
    public void setDifficulty(Difficulty difficulty) {
        EnumDifficulty diff = EnumDifficulty.getDifficultyEnum(difficulty.getId());
        this.wrapped.getWorldInfo().setDifficulty(diff);
        for (Player player : this.players) {
            ((NailedPlayer) player).sendPacket(new S41PacketServerDifficulty(diff, false));
        }
        if(difficulty == Difficulty.PEACEFUL){
            wrapped.setAllowedSpawnTypes(false, true);
        }else{
            wrapped.setAllowedSpawnTypes(true, true);
        }
    }

    public WorldServer getWrapped() {
        return wrapped;
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("id", getDimensionId())
                .add("name", getName())
                .add("dimension", getDimension())
                .add("gameRules", gameRules)
                .toString();
    }
}
